File: algorithms/sort/radixSort.ts
export function radixSort(arr: number[], base: number) {
  function radixSort(arr: number[], baseNow: number) {
    let maxBase: number = 0;
    arr.forEach((e) => {
      let thiBase = 10 ** (e.toString().length - 1);
      maxBase = maxBase < thiBase ? thiBase : maxBase;
    });
    let bucket: number[][];
    arr.forEach((e) => {
      Math.floor((e % baseNow) / Math.floor(baseNow / base));
    });
  }
}

radixSort([2, 1, 1000, 3], 10);
//第一位:  8  十进制:  8% (10**1) /(10**1/10)=8
//第二位   48 十进制: 48% (10**2) /(10**2/10)=4
// 8  -> 8进制的话  --> 0     8%8 =0

function getDigit(num: number, dig: number, base: number = 10): number {
  return Math.floor(
    (num % base ** (dig + 1)) / Math.floor(10 ** (dig + 1) / base),
  );
}
console.log(getDigit(1236, 2, 10));

File: algorithms/sort/radixSort.go
package sort

import (
	"math"
	"strconv"
)

type Bucket [16][]int

// 获取使用某进制下数组里出现的最大的位数，比如[1,2,3000,234,67] => 4
func getMaxRad(arr []int, base int) int {
	max := 0
	for _, v := range arr {
		l := len(strconv.FormatInt(int64(v), base))
		if max < l {
			max = l
		}
	}
	return max
}

// 把二维bucket展开
func flatBuk(bu Bucket) []int {
	res := []int{}
	for _, v := range bu {
		res = append(res, v...)
	}
	return res
}

// 拿到这个数某进制下某位的数字 0代表各位 如果 256 ，16进制,1 相当于拿到的0xFF 的十位：F 即 15
func GetDigit(num int, dig int, base int) int {
	times := math.Pow(float64(base), float64(dig+1))
	remain := num % int(times)
	return remain / int(math.Pow(float64(base), float64(dig)))
}

func RSort(arr []int, base int) []int {
	maxDig := getMaxRad(arr, base)
	// fmt.Println("maxDig", maxDig)

	return rdx(arr, 0, base, maxDig)
}

func RSort10(arr []int) []int {
	return RSort(arr, 10)
}

func rdx(ar []int, dig int, base int, maxDig int) []int {
	if dig >= maxDig {
		return ar
	}
	var bc Bucket

	for i := 0; i < len(ar); i++ {
		digitValue := GetDigit(ar[i], dig, base)
		bc[digitValue] = append(bc[digitValue], ar[i])
	}
	flat := flatBuk(bc)
	dig++
	return rdx(flat, dig, base, maxDig)
}
